framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,601 @@
+            }
+        }
+    }
+
+    /**
+     * For a given class, this prints the hierarchy of outer classes, and returns the number of
+     * curly braces to close with. The classes are printed with appropriate opening curly braces, in
+     * standard Java style. This routine does not attempt to indent them correctly.
+     *
+     * <p>When an inner class is present in an AScene, its name is something like "Outer.Inner".
+     * Writing a stub file with that name would be useless to the stub parser, which expects inner
+     * classes to be properly nested.
+     *
+     * @param basename the simple name of the class (without the package), in fully-qualified form
+     *     (that is, without any {@code $}s)
+     * @param classname the fully-qualified name of the class in question, so that it can be printed
+     *     as an enum if it is one
+     * @param aClass the AClass for {@code classname}
+     * @param printWriter the writer where the class definition should be printed
+     * @return the number of outer classes within which this class is nested
+     */
+    private int printClassDefinitions(
+            String basename,
+            @FullyQualifiedName String classname,
+            AClass aClass,
+            PrintWriter printWriter) {
+
+        String nameToPrint = basename;
+        String remainingInnerClassNames = "";
+        if (basename.contains(".")) {
+            nameToPrint = basename.substring(0, basename.indexOf('.'));
+            remainingInnerClassNames = basename.substring(basename.indexOf('.') + 1);
+        }
+
+        // For any outer class, print "class".  For a leaf class, print "enum" or "class".
+        if ("".equals(remainingInnerClassNames) && enumConstants.containsKey(classname)) {
+            printWriter.print("enum ");
+        } else {
+            printWriter.print("class ");
+        }
+        formatAnnotations(aClass.tlAnnotationsHere);
+        printWriter.print(nameToPrint);
+        printTypeParameters(classname, printWriter);
+        printWriter.println(" {");
+        if ("".equals(remainingInnerClassNames)) {
+            return 0;
+        } else {
+            return 1
+                    + printClassDefinitions(
+                            remainingInnerClassNames, classname, aClass, printWriter);
+        }
+    }
+
+    /**
+     * Prints all the fields of a given class
+     *
+     * @param aClass the class whose fields should be printed
+     * @param fullyQualifiedClassname the fully-qualified name of the class
+     * @param printWriter the writer on which to print the fields
+     */
+    private void printFields(
+            AClass aClass,
+            @FullyQualifiedName String fullyQualifiedClassname,
+            PrintWriter printWriter) {
+        for (Map.Entry<String, AField> fieldEntry : aClass.fields.entrySet()) {
+            String fieldName = fieldEntry.getKey();
+            AField aField = fieldEntry.getValue();
+            printWriter.println();
```

Comment:
It is standard to add a line separator at the end of a method, and that is how other methods like `printMethodSignature` (just below) work.  So this convention is surprising.  I suggest changing it, or if you don't do that, document the convention about newlines.  (It's probably easier and clearer to change it, though.)